File: C++/0402.cpp
#include <string>

class Solution {
public:
    std::string removeKdigits(std::string num, int k) {
        std::string stack = "";
        for (char c : num) {
            while (k && stack.size() && stack.back() > c) {
                stack.pop_back();
                k--;
            }
            if(stack.size() || c != '0') stack.push_back(c);
        }
        
        while(stack.size() && k--)
            stack.pop_back();
        
        return stack.size() == 0 ? "0" : stack;
    }
};

// Topic: Stack
// Time Complexity: O(n)
// Space Complexity: O(n)

File: Python/0402.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        stack = []
        for char in num:
            while k > 0 and stack and stack[-1] > char:
                k -= 1
                stack.pop()
            if(stack or char != "0"): stack.append(char)
            
        stack = stack[:len(stack)-k]

        ans = "".join(stack)

        return ans if ans else "0"
            
